src/io/asyncsocket.c

```diff
@@ -497,6 +526,102 @@ static void push_name_and_port(MVMThreadContext *tc, struct sockaddr_storage *na
     MVM_repr_push_o(tc, arr, port_o);
 }
 
+typedef struct {
+    uv_tcp_t         *socket;
+    MVMThreadContext *tc;
+    int               work_idx;
+} SocketInfo;
+
+static void socket_setup(MVMThreadContext *tc, uv_loop_t *loop, MVMObject *async_task, void *data) {
+    int r;
+
+    /* Add to work in progress. */
+    MVMObject    *arr = MVM_repr_alloc_init(tc, tc->instance->boot_types.BOOTArray);
+    MVMAsyncTask *t   = (MVMAsyncTask *)async_task;
+    SocketInfo   *si  = (SocketInfo *)data;
+    si->tc            = tc;
+    si->work_idx      = MVM_io_eventloop_add_active_work(tc, async_task);
+
+    /* Create and initialize socket. */
+    si->socket       = MVM_malloc(sizeof(uv_tcp_t));
+    si->socket->data = data;
+    if ((r = uv_tcp_init(loop, si->socket)) < 0) {
+        MVM_repr_push_o(tc, arr, t->body.schedulee);
+        MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTIO);
+        MVMROOT(tc, arr, {
+            MVMString *msg_str = MVM_string_ascii_decode_nt(tc, tc->instance->VMString, uv_strerror(r));
+            MVMObject *msg_box = MVM_repr_box_str(tc, tc->instance->boot_types.BOOTStr, msg_str);
+            MVM_repr_push_o(tc, arr, msg_box);
+        });
+        MVM_repr_push_o(tc, t->body.queue, arr);
+
+        /* Cleanup handles. */
+        uv_close((uv_handle_t *)si->socket, free_on_close_cb);
+        si->socket = NULL;
+        MVM_io_eventloop_remove_active_work(tc, &si->work_idx);
+    } else {
+        MVMROOT2(tc, arr, t, {
```

Comment:
Why `MVMROOT` `t` here but not in the above conditional branch?

Reply:
Missed that